stan/math/prim/err/is_column_index.hpp

```diff
@@ -11,15 +11,16 @@ namespace math {
  * Return <code>true</code> if column index is in bounds.
  * By default this is a 1-indexed check (as opposed to zero-indexed).
  * Behavior can be changed by setting <code>stan::error_index::value</code>.
- * @tparam EigMat A type derived from `EigenBase` with dynamic rows and columns
+ * @tparam EigMat A type derived from `EigenBase`
  * @param y matrix to test
  * @param i column index to check
  * @return <code>true</code> if column index is in bounds
  */
-template <typename EigMat, require_eigen_matrix_t<EigMat>* = nullptr>
+template <typename EigMat, require_eigen_t<EigMat>* = nullptr>
 inline bool is_column_index(const EigMat& y, size_t i) {
   return i >= stan::error_index::value
-         && i < static_cast<size_t>(y.cols()) + stan::error_index::value;
+         && i < static_cast<size_t>(to_ref(y).cols())
```

Comment:
You are making a copy just to see how many columns it has. Why?

Reply:
Ah good catch, I'll give all the similar error functions a look over for this as well

Reply:
One thing I'm also thinking about. I'd guess 9/10 times someone passing an expression to the error checking functions they forgot to use a `to_ref()` in the higher level code. We could have a static_assert in these like

```
static_assert(!is_plain_type<T>::value, "Eigen Expression passed to {error_function_name} but eigen expressions must be evaluated at the call site before calling error functions.");

Reply:
That should go into its own PR. Also check should accept trivial expressions, not just plain types.

Reply:
This one is not fixed yet.

Reply:
true my b